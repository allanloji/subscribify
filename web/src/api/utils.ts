import { API_URL } from "@/utils/constants";
import { Api } from "./Api";

const api = new Api();
api.baseUrl = API_URL;

export { api };
