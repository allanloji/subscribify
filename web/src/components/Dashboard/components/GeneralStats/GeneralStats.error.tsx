import { Spacer } from "@/components/ui";
import StatCard from "../StatCard";
import { Frown } from "lucide-react";
import * as S from "./GeneralStats.styles";

function GeneralStatsError() {
  return (
    <>
      <h2>Overall Stats</h2>
      <Spacer />
      <S.Container>
        <StatCard
          title="Could not load stats"
          value="Error"
          icon={<Frown />}
          color="#D04848"
        />
      </S.Container>
    </>
  );
}

export default GeneralStatsError;
