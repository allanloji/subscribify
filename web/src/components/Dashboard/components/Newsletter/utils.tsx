const backgrounds = [
  {
    primary: "#e1f0da",
    secondary: "#d4e7c5",
  },
  {
    primary: "#f6fedb",
    secondary: "#e6d3a3",
  },
  {
    primary: "#f9e3e3",
    secondary: "#f0c3c3",
  },
  {
    primary: "#e3e3f9",
    secondary: "#c3c3f0",
  },
  {
    primary: "#e3f9f9",
    secondary: "#c3f0f0",
  },
  {
    primary: "#f9e3f9",
    secondary: "#f0c3f0",
  },
  {
    primary: "#f9f3e3",
    secondary: "#f0d3c3",
  },
  {
    primary: "#f3f9e3",
    secondary: "#d3f0c3",
  },
  {
    primary: "#f3e3f9",
    secondary: "#d3c3f0",
  },
  {
    primary: "#e3f9e3",
    secondary: "#c3f0c3",
  },
  {
    primary: "#f9f3f3",
    secondary: "#f0d3d3",
  },
  {
    primary: "#98c1d9",
    secondary: "#6e9ebf",
  },
];

export const getBackground = (id: string) => {
  const sanitizedUuid = id.replace(/-/g, "");

  const uuidNumber = parseInt(sanitizedUuid.substring(0, 8), 16);

  const index = (uuidNumber * 10) % backgrounds.length;
  console.log(id, uuidNumber, index);

  return backgrounds[index];
};
