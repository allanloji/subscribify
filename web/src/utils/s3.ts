import { v4 as uuid } from "uuid";
import {
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

const s3 = new S3Client({
  region: process.env.NEXT_PUBLIC_AWS_REGION || "",
  credentials: {
    accessKeyId: process.env.NEXT_PUBLIC_AWS_ACCESS_KEY_ID || "",
    secretAccessKey: process.env.NEXT_PUBLIC_AWS_SECRET_ACCESS_KEY || "",
  },
});

export async function uploadFile(file: File, name?: string) {
  const key = `${uuid()}.${file.name.split(".").pop()}`;
  const command = new PutObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: name || key,
    Body: file,
  });

  try {
    await s3.send(command);
    return key;
  } catch (err) {
    console.error(err);
  }
}

export async function getFile(key: string) {
  const command = new GetObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: key,
  });

  const url = await getSignedUrl(s3, command, { expiresIn: 3600 });

  return url;
}
