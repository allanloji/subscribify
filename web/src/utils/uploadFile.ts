import { v4 as uuid } from "uuid";
import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";

const s3 = new S3Client({
  region: process.env.NEXT_PUBLIC_AWS_REGION || "",
  credentials: {
    accessKeyId: process.env.NEXT_PUBLIC_AWS_ACCESS_KEY_ID || "",
    secretAccessKey: process.env.NEXT_PUBLIC_AWS_SECRET_ACCESS_KEY || "",
  },
});

export async function uploadFile(file: File) {
  const key = `${uuid()}.${file.name.split(".").pop()}`;
  const command = new PutObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: key,
    Body: file,
  });

  try {
    await s3.send(command);
    return key;
  } catch (err) {
    console.error(err);
  }
}
