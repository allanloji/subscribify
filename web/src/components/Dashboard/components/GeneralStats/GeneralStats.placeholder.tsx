import { Spacer } from "@/components/ui";
import Skeleton from "react-loading-skeleton";
import "react-loading-skeleton/dist/skeleton.css";
import StatCard from "../StatCard";
import * as S from "./GeneralStats.styles";

function GeneralStatsPlaceholder() {
  return (
    <>
      <Skeleton height={24} width={300} />
      <Spacer />
      <S.Container>
        <StatCard.Placeholder />
        <StatCard.Placeholder />
        <StatCard.Placeholder />
      </S.Container>
    </>
  );
}

export default GeneralStatsPlaceholder;
