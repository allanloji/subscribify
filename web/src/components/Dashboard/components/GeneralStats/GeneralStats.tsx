import { Spacer } from "@/components/ui";
import { useQuery } from "@tanstack/react-query";
import StatCard from "../StatCard";
import { MailCheck, UserX, Users } from "lucide-react";
import * as S from "./GeneralStats.styles";
import GeneralStatsPlaceholder from "./GeneralStats.placeholder";
import GeneralStatsError from "./GeneralStats.error";
import { queries } from "@/api/queries";

function GeneralStats() {
  const {
    data: stats,
    isLoading: isLoadingStats,
    isError: isErrorStats,
  } = useQuery({ ...queries.stats.all });

  if (isLoadingStats) {
    return <GeneralStatsPlaceholder />;
  }

  if (isErrorStats || !stats) {
    return <GeneralStatsError />;
  }

  return (
    <section>
      <h2>Overall Stats</h2>
      <Spacer />
      <S.Container>
        <StatCard
          title="Total recipients"
          value={stats.totalRecipients}
          icon={<Users />}
          color="#D04848"
        />
        <StatCard
          title="Total emails sent"
          value={stats.totalEmailsSent}
          icon={<MailCheck />}
          color="#6895D2"
        />
        <StatCard
          title="Total unsubscribes"
          value={stats.totalUnsubscribes}
          icon={<UserX />}
          color="#F3B95F"
        />
      </S.Container>
    </section>
  );
}

export default GeneralStats;
