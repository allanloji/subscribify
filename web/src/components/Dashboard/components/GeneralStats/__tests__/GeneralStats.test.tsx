import { beforeAll, afterEach, afterAll, expect, test } from "vitest";
import { server } from "@/utils/tests/server";
import { waitFor } from "@testing-library/react";
import GeneralStats from "..";
import { render } from "@/utils/tests/renderApp";
import { HttpResponse, http } from "msw";
import { API_URL } from "@/utils/constants";

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("renders the general stats placeholder", async () => {
  const { container } = render(<GeneralStats />);

  expect(container).toMatchSnapshot();
});

test("renders the general stats with data", async () => {
  const { queryByText } = render(<GeneralStats />);

  await waitFor(() => {
    expect(queryByText("Total recipients")).not.toBe(null);
    expect(queryByText("Total emails sent")).not.toBe(null);
    expect(queryByText("Total unsubscribes")).not.toBe(null);
  });
});

test("renders the general stats error", async () => {
  server.use(http.get(`${API_URL}/statistics`, () => HttpResponse.error()));

  const { queryByText } = render(<GeneralStats />);

  await waitFor(() => {
    expect(queryByText("Could not load stats")).not.toBe(null);
  });
});
