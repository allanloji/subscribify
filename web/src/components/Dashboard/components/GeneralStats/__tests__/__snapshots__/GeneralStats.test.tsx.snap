// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the general stats placeholder 1`] = `
<div>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 300px; height: 24px;"
    >
      ‌
    </span>
    <br />
  </span>
  <div
    class="css-g1gnh4"
  />
  <div
    class="css-1bhr2z7"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="width: 300px; height: 100px; border-radius: 0.75rem;"
      >
        ‌
      </span>
      <br />
    </span>
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="width: 300px; height: 100px; border-radius: 0.75rem;"
      >
        ‌
      </span>
      <br />
    </span>
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="width: 300px; height: 100px; border-radius: 0.75rem;"
      >
        ‌
      </span>
      <br />
    </span>
  </div>
</div>
`;
